### PR TITLE
We need to fix labels if the user requests on volumes

### PR DESCRIPTION
**Pull request by rhatdan**
Currently local volumes and other volumes that support SELinux do
not get labeled correctly.  This patch will allow a user to specify
:Z or :z when  mounting a volume and have it fix the label of the newly
created volume.

Signed-off-by: Dan Walsh <dwalsh@redhat.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
Fixes this bugzilla https://bugzilla.redhat.com/show_bug.cgi?id=1401260

**Comment by rhatdan** [comment]
I did testing on docker-1.12 branch, I have not gotten 1.14 to work yet.

**Comment by runcom** [comment]
LGTM (the bugzilla isn't mentioning that this issue happends with `--selinux-enabled`, btw, I can reproduce it and this patch fixes it).

**Comment by cpuguy83** [comment]
Not sure I follow that issue... but isn't this the same as https://github.com/docker/docker/blob/master/volume/volume.go#L155-L159

I guess the relabel just needs to be moved up before the `if m.Volume != nil`

**Comment by rhatdan** [comment]
I think you would need it in both cases.  Since the second case the source directory is being created and needs to be labeled.

**Comment by rhatdan** [comment]
The Relabel has to happen after this code also.
```
if bind.Driver == volume.DefaultDriverName {
				setBindModeIfNull(bind)
}
```
Since this says if "local" volume and bind.Mode is Null, then set it to "z"

**Comment by cpuguy83** [comment]
This `Setup()` function is called when the container is started, which seems like a good place, we just aren't setting it for volumes, only binds.

**Comment by rhatdan** [comment]
@cpuguy83 Do you like this one better?  Should I relabel before or after the mount call?

**Comment by cpuguy83** [comment]
How about a defer at the top of the function... we'll have to name the returns so we can check if err is nil in the defer, but since there's no happy case where we don't want to relabel, I think it's safer.

**Comment by rhatdan** [comment]
@cpuguy83 Kind of ugly, is this what  you had in mind?  Any cleaner way to do this?

File: volume/volume.go [hunk 1]
```diff
@@ -125,20 +125,38 @@ type MountPoint struct {
 // Setup sets up a mount point by either mounting the volume if it is
 // configured, or creating the source directory if supplied.
 func (m *MountPoint) Setup(mountLabel string, rootUID, rootGID int) (string, error) {
+	var (
+		err  error
```
Comment (cpuguy83) [inline volume/volume.go]: declar this in the return instead of here.

File: volume/volume.go [hunk 2]
```diff
@@ -125,20 +125,38 @@ type MountPoint struct {
...
+		path string
+	)
+	defer func() (string, error) {
```
Comment (cpuguy83) [inline volume/volume.go]: no need to declare a return here

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

**Comment by rhatdan** [comment]
@cpuguy83 PTAL

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Comment by LK4D4** [comment]
Code is kinda weird, but I'm going to merge it anyway.

**Review by LK4D4 — APPROVED** [review]
LGTM

**Pull request merged**